indexer/pkg/discovery/message_discovery.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/smartcontractkit/chainlink-ccv/protocol"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+var _ common.MessageDiscovery = (*AggregatorMessageDiscovery)(nil)
+
+type AggregatorMessageDiscovery struct {
+	logger           logger.Logger
+	config           Config
+	aggregatorReader *readers.ResilientReader
+	storageSink      common.IndexerStorage
+	monitoring       common.IndexerMonitoring
+	messageCh        chan protocol.CCVData
+	stopCh           chan struct{}
+	doneCh           chan struct{}
+	readerLock       *sync.Mutex
+}
+
+type Config struct {
+	PollInterval       time.Duration
+	Timeout            time.Duration
+	MessageChannelSize int
+}
+
+type Option func(*AggregatorMessageDiscovery)
+
+func WithAggregator(aggregator *readers.ResilientReader) Option {
+	return func(a *AggregatorMessageDiscovery) {
+		a.aggregatorReader = aggregator
+	}
+}
+
+func WithMonitoring(monitoring common.IndexerMonitoring) Option {
+	return func(a *AggregatorMessageDiscovery) {
+		a.monitoring = monitoring
+	}
+}
+
+func WithStorage(storage common.IndexerStorage) Option {
+	return func(a *AggregatorMessageDiscovery) {
+		a.storageSink = storage
+	}
+}
+
+func WithLogger(lggr logger.Logger) Option {
+	return func(a *AggregatorMessageDiscovery) {
+		a.logger = lggr
+	}
+}
+
+func WithConfig(config Config) Option {
+	return func(a *AggregatorMessageDiscovery) {
+		a.config = config
+	}
+}
+
+func NewAggregatorMessageDiscovery(opts ...Option) (common.MessageDiscovery, error) {
+	a := &AggregatorMessageDiscovery{
+		stopCh:     make(chan struct{}),
+		doneCh:     make(chan struct{}),
+		readerLock: &sync.Mutex{},
+	}
+
+	// Apply all options
+	for _, opt := range opts {
+		opt(a)
+	}
+
+	// Create the buffered message channel
+	a.messageCh = make(chan protocol.CCVData, a.config.MessageChannelSize)
+
+	// Validata the configuration
+	if err := a.validate(); err != nil {
+		return nil, err
+	}
+
+	return a, nil
+}
+
+func (a *AggregatorMessageDiscovery) validate() error {
+	if a.config.MessageChannelSize <= 0 {
+		return errors.New("invalid message channel size")
+	}
+
+	if a.config.PollInterval <= 0 {
+		return errors.New("invalid poll interval")
+	}
+
+	if a.config.Timeout <= a.config.PollInterval {
+		return errors.New("invalid timeout, needs to be greater then poll interval")
+	}
+
+	if a.aggregatorReader == nil {
+		return errors.New("aggregator must be specified")
+	}
+
+	if a.logger == nil {
+		return errors.New("logger must be specified")
+	}
+
+	if a.monitoring == nil {
+		return errors.New("monitoring must be specified")
+	}
+
+	if a.storageSink == nil {
+		return errors.New("storage must be specified")
+	}
+
+	return nil
+}
+
+func (a *AggregatorMessageDiscovery) Start(ctx context.Context) chan protocol.CCVData {
+	go a.run(ctx)
+	a.logger.Info("MessageDiscovery Started")
+
+	// Return a channel that emits all messages discovered from the aggregator
+	return a.messageCh
```

Comment:
nit: Will we want multiple subscriber at the same time for the messageIds? Or perhaps we will create one message discovery by subscriber?